ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/Jumping.hs

```diff
@@ -0,0 +1,656 @@
+            let fragment = jTheirFragment jumpInfo
+            modifyTVar (cschState (handle context)) $ \csState -> csState {csCandidate = fragment}
+            writeTVar (cschJumpInfo (handle context)) $ Just jumpInfo
+            case jumperState of
+              LookingForIntersection badJumpInfo ->
+                -- @AF.headPoint fragment@ is in @badFragment@, as the jumper
+                -- looking for an intersection is the only client asking for its
+                -- jumps.
+                lookForIntersection nextJumpVar (AF.headPoint fragment) badJumpInfo
+              Happy ->
+                writeTVar (cschJumping (handle context)) $
+                  Jumper nextJumpVar (AF.headPoint fragment) Happy
+              FoundIntersection{} ->
+                -- Only happy jumpers are asked to jump by the dynamo, and only
+                -- jumpers looking for an intersection are asked to jump by
+                -- themselves.
+                error "processJumpResult: Jumpers in state FoundIntersection shouldn't be further jumping."
+            pure (Just jumpInfo)
+
+          RejectedJump badJumpInfo -> do
+            -- @goodPoint@ is in @jTheirFragment jumpInfo@. If the jump was
+            -- requested by the dynamo, this holds because the dynamo is not
+            -- allowed to rollback before the jumps that it requests.
+            --
+            -- If the jump was requested by the jumper, this holds because the
+            -- jumper is looking for an intersection, and such jumper only asks
+            -- for jumps that meet this condition.
+            lookForIntersection nextJumpVar goodPoint badJumpInfo
+            pure Nothing
+  where
+    -- Avoid redundant constraint "HasHeader blk" reported by some ghc's
+    _ = getHeaderFields @blk
+    -- | Given a good point (where we know we agree with the dynamo) and a bad
+    -- fragment (where we know the tip disagrees with the dynamo), either decide
+    -- that we know the intersection for sure (if the bad point is the successor
+    -- of the good point) or program a jump somewhere in the middle to refine
+    -- those points.
+    --
+    -- PRECONDITION: The good point is in the bad fragment.
+    lookForIntersection nextJumpVar goodPoint badJumpInfo = do
+      let badFragment = jTheirFragment badJumpInfo
+      case snd <$> AF.splitAfterPoint badFragment goodPoint of
+        Nothing ->
+          error "processJumpResult: goodPoint not in badFragment"
+        Just nextFragment -> do
+          let len = AF.length nextFragment
+          if len <= 1 then do
+            -- If the fragment only contains the bad tip, we know the
+            -- intersection is the good point.
+            writeTVar (cschJumping (handle context)) $
+              Jumper nextJumpVar goodPoint $
+                FoundIntersection $ AF.headPoint badFragment
+            demoteObjector (stripContext context)
+            electNewObjector (stripContext context)
```

Comment:
Looking at the code, I think electing the same objector is as if it never had been demoted.

Reply:
There is still the change that `electNewObjector` finds another objector that forks from the same point as the current one, so then it would have been better to just keep the current one. This might in particular occur if the objector is honest, because we expect many other honest peers to object and they will all object with the same intersection with the dynamo.

Reply:
I tried to address that in 7cee943.